planner/core/push_down_predicate.go

```diff
@@ -0,0 +1,246 @@
+// Copyright 2023 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package core
+
+import (
+	"sort"
+
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/kv"
+	"github.com/pingcap/tidb/parser/ast"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/logutil"
+	"go.uber.org/zap"
+)
+
+const (
+	// selectivity = (row count after filter) / (row count before filter), smaller is better
+	selectivityThreshold = 0.7
+	// income = (1 - selectivity) * resetColumnCount, greater is better
+	incomeImproveThreshold = 0.1
+)
+
+// expressionGroup is used to store
+// 1. a group of expressions
+// 2. the selectivity of the expressions
+type expressionGroup struct {
+	exprs       []expression.Expression
+	selectivity float64
+}
+
+func pushDownPredicate(sctx sessionctx.Context, plan PhysicalPlan) {
```

Comment:
IMO the function name may conflict with other stuff in optimizer, when I look at the function, I think this function just push down filter as low as possible

Reply:
`pushDownPredicate  -> predicatePushDownToTableScan`

`predicatePushDownToTableScan -> predicatePushDownToTableScanImpl`